### ISSUE TITLE
Password reset should invalidate active sessions, and requires other improvements

### ISSUE THREAD

**Issue by WeaverThree** [HabitRPG/habitica#4499]
I went to login in Firefox to use a stylish theme, and realized I'd forgotten my new password already. Then I wanted to check something without the theme, and figured the easiest way would be to go back to Chrome and log in again.

(To be clear, I didn't do any part of the password reset operation in Chrome except copying text out of gmail.)

Except I didn't have to log in again, the chrome session was still valid.

Now if I were to have changed my password due to suspecting someone using my account without authorization, I would want all old sessions tokens to be revoked at that point. I mean, it's not terribly likely with this site, but that's how password security is usually supposed to work... right?
##

**Comment by lefnire** [comment]
absolutely. There's a checklist of proper auth-handling we'll need to move too soon as we can:
- [ ] oauth. our current static uuid/api-token setup is cludgey - should be fixed as part of ifttt integration  / oath2 authentication improvements - https://github.com/HabitRPG/habitrpg/tree/oauth2-2015-01-10
- [ ] resettable api tokens - should be fixed as part of ifttt integration / oath2 authentication improvements
- [x] password reset emails with one-time link, rather than sending a new email via password
- [ ] email verification on signup
- [ ] sign-outs of other devices when password reset (as you mentioned)

**Comment by crookedneighbor** [comment]
@lefnire Converted your ordered list to a checklist.

**Comment by lefnire** [comment]
- [x] √

**Comment by Alys** [comment]
I added mention of ifttt integration, which should fix the first two items just as soon as I get right onto that.

**Comment by lefnire** [comment]
Our accelerator guys have recommended not using email verification on registration, but instead using a 3rd party email validation tool (simply to test that it's a legit email, not to require them to verify from their inbox). They say it's a barrier-to-entry and doesn't provide any real service, and the only downside is users signing up with someone else's email address (which.. like... why). Open to discussion, just thought I'd throw this in there. 'Course, we _definitely_ want password resets to be a one-time link in inbox, diff story.

**Comment by Alys** [comment]
One thing to note is that a password reset feature that logs out the user on another device can be used as a denial-of-service attack against the owner of the account. To prevent that, we'd probably want to limit the number of password resets per day to maybe three (or a few more?), and have the count reset by cron.

**Comment by lefnire** [comment]
TMK, we'd want to send an email to the registered user with a link asking them reset the password by clicking the link. They click the link, enter the new password, click submit - and only then does it log them out from other devices. Ie, they have to own the email address in the first place to get the process started, so wouldn't be susceptible to DoS (unless I'm missing something?)

**Comment by Alys** [comment]
@lefnire Ha, yes, that would work well. :)

**Comment by WeaverThree** [comment]
@lefnire Personally, I don't really trust a static email address validator to not throw false negatives on unusual but completely valid addresses. References and discussion along these lines:

https://stackoverflow.com/questions/46155/validate-email-address-in-javascript
http://ex-parrot.com/~pdw/Mail-RFC822-Address.html
http://www.regular-expressions.info/email.html

Takeaway: RFC822 is complicated, yo o.o

As a user, I'd really rather the system send me some kind of confirmation email so I know that I've gotten things set up right; plus then I can search my email history to see if I ever actually set up an account somewhere.

Everything else you guys are doing looks great, though. :3

**Comment by lemoness** [comment]
@starweaver for what it's worth, we do send a welcome email now!

**Comment by Alys** [comment]
We could use an email address validator and then if that showed the address was invalid, give the user an option to either edit their address or have an email sent to them that they must reply to to prove it's valid before their registration would be complete.

**Comment by paglias** [comment]
We could just see if the welcome email went through or bounced back
Il 15/feb/2015 22:46 "Alice Harris" notifications@github.com ha scritto:

**Comment by Alys** [comment]
You can't rely on bounces. Mail servers can silently drop messages.

Is the point of address validation to ensure that all registered users have valid addresses? If so, it needs to be done before account creation, unfortunately. Although they can always specify disposable addresses, so this isn't a perfect protection against trolls.

**Comment by lemoness** [comment]
Still, it's better than nothing - and also makes it easier to reach users
via email.

**Comment by lemoness** [comment]
That's not to say that we need to use one of the services already proposed,
but it would definitely be nice to find some way to do this in a
satisfactory manner.

**Comment by Alys** [comment]
From several comments back:
_"the only downside is users signing up with someone else's email address (which.. like... why)."_

I've seen at least two complaints of that in emails to the admin@habitica.com address. I'm certain that one of them was a legitimate complaint. The other one felt a bit more like the user just forgot that they'd signed up ages ago, but I might be misjudging that.

Unrelated to that, we had a (very polite!) complaint today from a teacher who had one of his students reset the teacher's password. No significant harm done since the reset email went to the teacher, but as the teacher pointed out it's "a minor annoyance that could grow louder." If the entire class starts thinking that would be a fun thing to do.......

**Comment by librarianmage** [comment]
What is the status on this?

**Comment by paglias** [comment]
@MathWhiz we want the actual password reset to happen only with the user authorization (so after they click on a link in the email and not as soon as the email is sent) and if you want to work on this, go ahead. What we cannot do, with the current auth system, is invalidating old sessions, that'll have to wait

**Comment by saraolson** [comment]
changing your password now resets your API token and logs you out of other platforms/browsers/tabs
